File: solutions/Java/[0039_Medium] Combination Sum/combinationSum.java
class Solution {
  private List<List<Integer>> result = new ArrayList<>();

  public List<List<Integer>> combinationSum(int[] candidates, int target) {
    backtrackCombination(0, target, candidates, new ArrayList<>());
    return result;
  }

  public void backtrackCombination(int start, int target, int[] candidates, List<Integer> combination) {
    if (target < 0) {
      return;
    }

    if (target == 0) {
      result.add(new ArrayList(combination));
      return;
    }

    for (int i = start; i < candidates.length; i++) {
      combination.add(candidates[i]);
      backtrackCombination(i, target - candidates[i], candidates, combination);
      combination.remove(combination.size() - 1);
    }
  }
}

File: solutions/TypeScript/[0039_Medium] Combination Sum/combinationSum.ts
function combinationSum(candidates: number[], target: number): number[][] {
  const result: number[][] = [];

  function backtrackCombination(
    start: number, balance: number, nums: number[], combination: number[]
  ) {
    if (balance < 0) {
      return;
    }

    if (balance === 0) {
      result.push([...combination]);
      return;
    }

    for (let i = start; i < nums.length; i++) {
      combination.push(nums[i]);
      backtrackCombination(i, balance - nums[i], nums, combination);
      combination.pop();
    }
  }

  backtrackCombination(0, target, candidates, []);
  return result;
}
